{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9759656c-716f-4bbc-8fe6-55bb22b19aaa",
   "metadata": {},
   "source": [
    "# Building a boolean network from a knowledge graph\n",
    "\n",
    "Use SigNOR\n",
    "\n",
    "Given a list of genes for which we want to construct a BN:\n",
    "\n",
    "1. Find the Steiner subgraph for the given list of genes.\n",
    "2. For each node in the subgraph, find all if its input nodes (i.e. all edges leading into that node).\n",
    "3. For each such node, there is an activating relation if the edge is \"up-regulates\", and a repressing relation if the edge is \"down-regulates\".\n",
    "4. Combine all of the edges with a \"joiner function\" - AND, OR, repressor wins, etc.\n",
    "\n",
    "Or, given an existing model (with some number of genes), a list of nodes, and a knowledge graph, how do we incorporate the new nodes into the model?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e24e0fa2-24ba-4079-8f09-35b70ed63c66",
   "metadata": {},
   "outputs": [],
   "source": [
    "import BNMPy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "717d34f3-a011-4fe9-8563-311a0c094be7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of genes found: 3\n",
      "[7128, 595, 23499]\n"
     ]
    }
   ],
   "source": [
    "bn_string, relations = BNMPy.load_signor_network(['TNFAIP3', 'CCND1', 'MACF1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "079e289b-e060-4d7e-8342-d61d1b7608da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCND1 = (! GSK3B) # Scores: GSK3B_inhibit:0.783\n",
      "GSK3B = (GSK3B) & (! MACF1) # Scores: GSK3B_activate:0.2; MACF1_inhibit:0.436\n",
      "MACF1 = (! GSK3B) # Scores: GSK3B_inhibit:0.436\n",
      "TNFAIP3 = TNFAIP3\n",
      "TRAF6 = (! GSK3B) & (TRAF6) & (! TNFAIP3) # Scores: GSK3B_inhibit:0.48; TRAF6_activate:0.2; TNFAIP3_inhibit:0.701\n"
     ]
    }
   ],
   "source": [
    "print(bn_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e232c4f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applied score cutoff 0.5, filtered to 19342/40940 edges\n",
      "number of genes found: 3\n",
      "[7128, 595, 23499]\n",
      "CCND1 = (CTNNB1) # Scores: CTNNB1_activate:0.8\n",
      "CTNNB1 = (! SRC) # Scores: SRC_inhibit:0.76\n",
      "SRC = SRC\n",
      "TNFAIP3 = TNFAIP3\n",
      "TRAF6 = (SRC) & (! TNFAIP3) # Scores: SRC_activate:0.566; TNFAIP3_inhibit:0.701\n"
     ]
    }
   ],
   "source": [
    "bn_string, relations = BNMPy.load_signor_network(['TNFAIP3', 'CCND1', 'MACF1'], score_cutoff = 0.5)\n",
    "print(bn_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "66f6a37a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of genes found: 3\n",
      "[7128, 595, 23499]\n",
      "CCND1 = 0 # Scores: GSK3B_inhibit:0.783\n",
      "GSK3B = (GSK3B & !MACF1) # Scores: GSK3B_activate:0.2; MACF1_inhibit:0.436\n",
      "MACF1 = 0 # Scores: GSK3B_inhibit:0.436\n",
      "TNFAIP3 = TNFAIP3\n",
      "TRAF6 = (TRAF6 & !GSK3B & !TNFAIP3) # Scores: GSK3B_inhibit:0.48; TRAF6_activate:0.2; TNFAIP3_inhibit:0.701\n"
     ]
    }
   ],
   "source": [
    "bn_string, relations = BNMPy.load_signor_network(['TNFAIP3', 'CCND1', 'MACF1'], joiner='majority')\n",
    "print(bn_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a07a6270-45f0-4971-a80c-8429dd81e437",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of genes found: 6\n",
      "[3845, 2778, 7157, 4089, 1029, 54894]\n",
      "CDKN2A = (! MYC) # inhibit_MYC:0.765\n",
      "GNAS = (! MDM2) # inhibit_MDM2:0.395\n",
      "GSK3B = (GSK3B) # activate_GSK3B:0.2\n",
      "KRAS = (SRC) # activate_SRC:0.656\n",
      "MDM2 = (TP53) # activate_TP53:0.968\n",
      "MYC = (! GSK3B) & (! SMAD4) # inhibit_GSK3B:0.719; inhibit_SMAD4:0.638\n",
      "RNF43 = RNF43\n",
      "SMAD4 = (! GSK3B) # inhibit_GSK3B:0.397\n",
      "SRC = (GNAS) & (GSK3B) & (SRC) # activate_GNAS:0.506; activate_GSK3B:0.383; activate_SRC:0.2\n",
      "TP53 = (! MDM2) & (GSK3B) & (! SRC) & (! RNF43) # inhibit_MDM2:0.968; activate_GSK3B:0.727; inhibit_SRC:0.524; inhibit_RNF43:0.452\n"
     ]
    }
   ],
   "source": [
    "bn_string, relations = BNMPy.load_signor_network(['KRAS', 'GNAS', 'TP53', 'SMAD4', 'CDKN2A', 'RNF43'], joiner='&')\n",
    "print(bn_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dece2511",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of genes found: 6\n",
      "[3845, 2778, 7157, 4089, 1029, 54894]\n",
      "CDKN2A = (! MYC) # inhibit_MYC:0.765\n",
      "GNAS = GNAS\n",
      "KRAS = (SRC) # activate_SRC:0.656\n",
      "MAPK1 = MAPK1\n",
      "MYC = (MAPK1) & (! SMAD4) # activate_MAPK1:0.733; inhibit_SMAD4:0.638\n",
      "SMAD4 = (MAPK1) # activate_MAPK1:0.511\n",
      "SRC = (GNAS) # activate_GNAS:0.506\n",
      "TP53 = (! SRC) & (MAPK1) # inhibit_SRC:0.524; activate_MAPK1:0.777\n"
     ]
    }
   ],
   "source": [
    "bn_string, relations = BNMPy.load_signor_network(['KRAS', 'GNAS', 'TP53', 'SMAD4', 'CDKN2A', 'RNF43'], joiner='&', score_cutoff = 0.5)\n",
    "print(bn_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4fe4db66-e05b-43fe-87d2-10301a2b3991",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of genes found: 6\n",
      "[3845, 2778, 7157, 4089, 1029, 54894]\n",
      "CDKN2A = (! MYC) # Scores: MYC_inhibit:0.765\n",
      "GNAS = (! MDM2) # Scores: MDM2_inhibit:0.395\n",
      "GSK3B = (GSK3B) # Scores: GSK3B_activate:0.2\n",
      "KRAS = (SRC) # Scores: SRC_activate:0.656\n",
      "MDM2 = (TP53) # Scores: TP53_activate:0.968\n",
      "MYC = (! GSK3B) | (! SMAD4) # Scores: GSK3B_inhibit:0.719; SMAD4_inhibit:0.638\n",
      "RNF43 = RNF43\n",
      "SMAD4 = (! GSK3B) # Scores: GSK3B_inhibit:0.397\n",
      "SRC = (GNAS) | (GSK3B) | (SRC) # Scores: GNAS_activate:0.506; GSK3B_activate:0.383; SRC_activate:0.2\n",
      "TP53 = (! MDM2) | (GSK3B) | (! SRC) | (! RNF43) # Scores: MDM2_inhibit:0.968; GSK3B_activate:0.727; SRC_inhibit:0.524; RNF43_inhibit:0.452\n"
     ]
    }
   ],
   "source": [
    "bn_string, relations = BNMPy.load_signor_network(['KRAS', 'GNAS', 'TP53', 'SMAD4', 'CDKN2A', 'RNF43'], joiner='|')\n",
    "print(bn_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6ac30ce2-4852-4a9e-9fc7-d366b9c8f5fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of genes found: 6\n",
      "[3845, 2778, 7157, 4089, 1029, 54894]\n",
      "CDKN2A = !MYC # Scores: MYC_inhibit:0.765\n",
      "GNAS = !MDM2 # Scores: MDM2_inhibit:0.395\n",
      "GSK3B = GSK3B # Scores: GSK3B_activate:0.2\n",
      "KRAS = SRC # Scores: SRC_activate:0.656\n",
      "MDM2 = TP53 # Scores: TP53_activate:0.968\n",
      "MYC = (!GSK3B & !SMAD4) # Scores: GSK3B_inhibit:0.719; SMAD4_inhibit:0.638\n",
      "RNF43 = RNF43\n",
      "SMAD4 = !GSK3B # Scores: GSK3B_inhibit:0.397\n",
      "SRC = (GNAS | GSK3B | SRC) # Scores: GNAS_activate:0.506; GSK3B_activate:0.383; SRC_activate:0.2\n",
      "TP53 = (!MDM2 & !SRC & !RNF43) & GSK3B # Scores: MDM2_inhibit:0.968; GSK3B_activate:0.727; SRC_inhibit:0.524; RNF43_inhibit:0.452\n"
     ]
    }
   ],
   "source": [
    "bn_string, relations = BNMPy.load_signor_network(['KRAS', 'GNAS', 'TP53', 'SMAD4', 'CDKN2A', 'RNF43'],\n",
    "                                                            joiner='inhibitor_wins')\n",
    "print(bn_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "55a8aa62-2b43-4a3e-aa30-cc62de0d3120",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('MDM2', 'GNAS', 'inhibit', 0.395),\n",
       " ('TP53', 'MDM2', 'activate', 0.968),\n",
       " ('GSK3B', 'GSK3B', 'activate', 0.2),\n",
       " ('GNAS', 'SRC', 'activate', 0.506),\n",
       " ('GSK3B', 'SRC', 'activate', 0.383),\n",
       " ('SRC', 'SRC', 'activate', 0.2),\n",
       " ('MDM2', 'TP53', 'inhibit', 0.968),\n",
       " ('GSK3B', 'TP53', 'activate', 0.727),\n",
       " ('SRC', 'TP53', 'inhibit', 0.524),\n",
       " ('RNF43', 'TP53', 'inhibit', 0.452),\n",
       " ('GSK3B', 'MYC', 'inhibit', 0.719),\n",
       " ('SMAD4', 'MYC', 'inhibit', 0.638),\n",
       " ('MYC', 'CDKN2A', 'inhibit', 0.765),\n",
       " ('SRC', 'KRAS', 'activate', 0.656),\n",
       " ('GSK3B', 'SMAD4', 'inhibit', 0.397)]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "relations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "fc0942c5-5a20-4926-9f10-bb54c2ee9ff7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of genes found: 6\n",
      "[3845, 2778, 7157, 4089, 1029, 54894]\n",
      "CDKN2A = 0 # Scores: MYC_inhibit:0.765\n",
      "GNAS = 0 # Scores: MDM2_inhibit:0.395\n",
      "GSK3B = (GSK3B) # Scores: GSK3B_activate:0.2\n",
      "KRAS = (SRC) # Scores: SRC_activate:0.656\n",
      "MDM2 = (TP53) # Scores: TP53_activate:0.968\n",
      "MYC = 0 # Scores: GSK3B_inhibit:0.719; SMAD4_inhibit:0.638\n",
      "RNF43 = RNF43\n",
      "SMAD4 = 0 # Scores: GSK3B_inhibit:0.397\n",
      "SRC = (GNAS & !GSK3B & !SRC) | (!GNAS & GSK3B & !SRC) | (GNAS & GSK3B & !SRC) | (!GNAS & !GSK3B & SRC) | (GNAS & !GSK3B & SRC) | (!GNAS & GSK3B & SRC) | (GNAS & GSK3B & SRC) # Scores: GNAS_activate:0.506; GSK3B_activate:0.383; SRC_activate:0.2\n",
      "TP53 = (GSK3B & !MDM2 & !SRC & !RNF43) # Scores: MDM2_inhibit:0.968; GSK3B_activate:0.727; SRC_inhibit:0.524; RNF43_inhibit:0.452\n"
     ]
    }
   ],
   "source": [
    "bn_string, relations = BNMPy.load_signor_network(['KRAS', 'GNAS', 'TP53', 'SMAD4', 'CDKN2A', 'RNF43'],\n",
    "                                                            joiner='majority')\n",
    "print(bn_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "80d88361-e070-4b1d-ad46-b131e565a633",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of genes found: 6\n",
      "[3845, 2778, 7157, 4089, 1029, 54894]\n",
      "CDKN2A = (!MYC) # Scores: MYC_inhibit:0.765\n",
      "GNAS = (!MDM2) # Scores: MDM2_inhibit:0.395\n",
      "GSK3B = (!GSK3B) | (GSK3B) # Scores: GSK3B_activate:0.2\n",
      "KRAS = (!SRC) | (SRC) # Scores: SRC_activate:0.656\n",
      "MDM2 = (!TP53) | (TP53) # Scores: TP53_activate:0.968\n",
      "MYC = (!GSK3B & !SMAD4) # Scores: GSK3B_inhibit:0.719; SMAD4_inhibit:0.638\n",
      "RNF43 = RNF43\n",
      "SMAD4 = (!GSK3B) # Scores: GSK3B_inhibit:0.397\n",
      "SRC = (!GNAS & !GSK3B & !SRC) | (GNAS & !GSK3B & !SRC) | (!GNAS & GSK3B & !SRC) | (GNAS & GSK3B & !SRC) | (!GNAS & !GSK3B & SRC) | (GNAS & !GSK3B & SRC) | (!GNAS & GSK3B & SRC) | (GNAS & GSK3B & SRC) # Scores: GNAS_activate:0.506; GSK3B_activate:0.383; SRC_activate:0.2\n",
      "TP53 = (!GSK3B & !MDM2 & !SRC & !RNF43) | (GSK3B & !MDM2 & !SRC & !RNF43) | (GSK3B & MDM2 & !SRC & !RNF43) | (GSK3B & !MDM2 & SRC & !RNF43) | (GSK3B & !MDM2 & !SRC & RNF43) # Scores: MDM2_inhibit:0.968; GSK3B_activate:0.727; SRC_inhibit:0.524; RNF43_inhibit:0.452\n"
     ]
    }
   ],
   "source": [
    "bn_string, relations = BNMPy.load_signor_network(['KRAS', 'GNAS', 'TP53', 'SMAD4', 'CDKN2A', 'RNF43'],\n",
    "                                                            joiner='plurality')\n",
    "print(bn_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19020452-d295-4556-8281-db5618d21b6c",
   "metadata": {},
   "source": [
    "## Combining KG knowledge graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94a9f1cc-10a2-43a7-88f8-31108097a057",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No initial state provided, using a random initial state\n",
      "Network loaded successfully. There are 38 genes in the network.\n",
      "number of genes: 38\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "dict_keys(['EGF', 'HBEGF', 'IGF1', 'NRG1', 'PTEN', 'STK11', 'EGFR', 'ERBB4', 'IGF1R', 'ERBB2', 'JAK1', 'STAT3', 'IRS1', 'GRB2', 'KRAS', 'MAP3K1', 'RAF1', 'MAP2K4', 'MAP2K1', 'PIK3CA', 'MAPK8', 'MAPK3', 'PIP3', 'PDPK1', 'AKT1', 'PRKAA1', 'GSK3B', 'TSC1', 'RHEB', 'MTOR', 'RPS6KB1', 'BAD', 'CCND1', 'BCL2', 'ELK1', 'FOS', 'ELK4', 'SP1'])"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load the Vundavalli KG\n",
    "file = './input_files/Vundavilli2020_standardized.txt'\n",
    "network = BNMPy.load_network_from_file(file)\n",
    "genes = network.nodeDict.keys()\n",
    "print(f\"number of genes: {len(genes)}\")\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d9c4399-5d87-4ba9-953d-5c3bf12cc0a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applied score cutoff 0.5, filtered to 19342/40940 edges\n",
      "number of genes found: 37\n",
      "[1950, 1839, 3479, 3084, 5728, 6794, 1956, 2066, 3480, 2064, 3716, 6774, 3667, 2885, 3845, 4214, 5894, 6416, 5604, 5290, 5599, 5595, 5170, 207, 5562, 2932, 7248, 6008, 2475, 6198, 572, 595, 596, 2002, 2353, 2005, 5669]\n",
      "AKT1 = !PTEN & (PDPK1 | MTOR | PIK3CA) # Scores: PDPK1_activate:0.749; MTOR_activate:0.929; PTEN_inhibit:0.634; PIK3CA_activate:0.816\n",
      "BAD = (!MAPK8 & !AKT1 & !RAF1) # Scores: MAPK8_inhibit:0.686; AKT1_inhibit:0.823; RAF1_inhibit:0.66\n",
      "BCL2 = (!MAPK8 & !BAD) & MAPK3 # Scores: MAPK8_inhibit:0.581; BAD_inhibit:0.801; MAPK3_activate:0.559\n",
      "CCND1 = !GSK3B & STAT3 # Scores: GSK3B_inhibit:0.783; STAT3_activate:0.787\n",
      "EGF = EGF\n",
      "EGFR = !MAPK3 & (ERBB2 | EGF | HBEGF) # Scores: ERBB2_activate:0.614; MAPK3_inhibit:0.556; EGF_activate:0.949; HBEGF_activate:0.767\n",
      "ELK1 = (MAPK8 | MAPK3) # Scores: MAPK8_activate:0.512; MAPK3_activate:0.6\n",
      "ERBB2 = (EGFR | NRG1 | EGF) # Scores: EGFR_activate:0.614; NRG1_activate:0.791; EGF_activate:0.79\n",
      "ERBB4 = !MAPK3 & (ERBB2 | NRG1 | HBEGF) # Scores: ERBB2_activate:0.529; MAPK3_inhibit:0.532; NRG1_activate:0.903; HBEGF_activate:0.748\n",
      "FOS = MAPK3 # Scores: MAPK3_activate:0.716\n",
      "GRB2 = (IRS1 | ERBB2 | ERBB4) # Scores: IRS1_activate:0.803; ERBB2_activate:0.85; ERBB4_activate:0.83\n",
      "GSK3B = !AKT1 # Scores: AKT1_inhibit:0.793\n",
      "HBEGF = HBEGF\n",
      "IGF1 = IGF1\n",
      "IGF1R = IGF1 # Scores: IGF1_activate:0.956\n",
      "IRS1 = (!MAPK8 & !MTOR & !PTEN & !MAPK3 & !PIK3CA & !RPS6KB1) & (JAK1 | IGF1R) # Scores: MAPK8_inhibit:0.771; MTOR_inhibit:0.765; PTEN_inhibit:0.576; JAK1_activate:0.708; IGF1R_activate:0.868; MAPK3_inhibit:0.706; PIK3CA_inhibit:0.721; RPS6KB1_inhibit:0.787\n",
      "JAK1 = EGFR # Scores: EGFR_activate:0.631\n",
      "KRAS = KRAS\n",
      "MAP2K1 = (RAF1 | MAP3K1) # Scores: RAF1_activate:0.749; MAP3K1_activate:0.66\n",
      "MAP2K4 = !AKT1 & MAP3K1 # Scores: AKT1_inhibit:0.593; MAP3K1_activate:0.73\n",
      "MAP3K1 = MAP3K1\n",
      "MAPK3 = MAP2K1 # Scores: MAP2K1_activate:0.752\n",
      "MAPK8 = MAP2K4 # Scores: MAP2K4_activate:0.752\n",
      "MTOR = (!RPS6KB1 & !TSC1) & PIK3CA # Scores: PIK3CA_activate:0.547; RPS6KB1_inhibit:0.96; TSC1_inhibit:0.705\n",
      "NRG1 = NRG1\n",
      "PDPK1 = !RPS6KB1 # Scores: RPS6KB1_inhibit:0.728\n",
      "PIK3CA = !PTEN & (IRS1 | ERBB4 | KRAS) # Scores: IRS1_activate:0.721; PTEN_inhibit:0.73; ERBB4_activate:0.608; KRAS_activate:0.91\n",
      "PRKAA1 = STK11 # Scores: STK11_activate:0.599\n",
      "PTEN = !STK11 # Scores: STK11_inhibit:0.604\n",
      "RAF1 = (!AKT1 & !MAPK3) & (KRAS | MAP2K1) # Scores: AKT1_inhibit:0.7; MAPK3_inhibit:0.637; KRAS_activate:0.847; MAP2K1_activate:0.749\n",
      "RPS6KB1 = !PTEN & (PDPK1 | MTOR | MAPK3) # Scores: PDPK1_activate:0.728; MTOR_activate:0.96; PTEN_inhibit:0.531; MAPK3_activate:0.607\n",
      "STAT3 = (MAPK8 | EGFR | MTOR | JAK1 | MAPK3) # Scores: MAPK8_activate:0.567; EGFR_activate:0.879; MTOR_activate:0.746; JAK1_activate:0.799; MAPK3_activate:0.718\n",
      "STK11 = STK11\n",
      "TSC1 = TSC1\n"
     ]
    }
   ],
   "source": [
    "bn_string, relations = BNMPy.load_signor_network(genes, joiner='inhibitor_wins', score_cutoff=0.5)\n",
    "print(bn_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "70b23ac1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No initial state provided, using a random initial state\n",
      "Network loaded successfully. There are 34 genes in the network.\n",
      "Merging Method: Inhibitor Wins\n",
      "Total Genes in Merged Network: 38\n",
      "Number of Genes in Each Individual Model:\n",
      "  Model 1: 38 genes\n",
      "  Model 2: 34 genes\n",
      "Overlapping Genes: 34\n",
      "Overlapping Genes List: AKT1, BAD, BCL2, CCND1, EGF, EGFR, ELK1, ERBB2, ERBB4, FOS, GRB2, GSK3B, HBEGF, IGF1, IGF1R, IRS1, JAK1, KRAS, MAP2K1, MAP2K4, MAP3K1, MAPK3, MAPK8, MTOR, NRG1, PDPK1, PIK3CA, PRKAA1, PTEN, RAF1, RPS6KB1, STAT3, STK11, TSC1\n",
      "\n",
      "Gene: AKT1\n",
      "  Model 1 Function: PIP3\n",
      "  Model 2 Function: !PTEN & ( MTOR | PDPK1 | PIK3CA )\n",
      "  Merged Function: !PTEN & ( MTOR | PDPK1 | PIK3CA | PIP3 )\n",
      "\n",
      "Gene: BAD\n",
      "Warning: possible conflicts for ['AKT1'], keeping only inhibitor.\n",
      "  Model 1 Function: ! ( AKT1 | RPS6KB1 )\n",
      "  Model 2 Function: !AKT1 & !MAPK8 & !RAF1\n",
      "  Merged Function: !AKT1 & !MAPK8 & !RAF1 & RPS6KB1\n",
      "\n",
      "Gene: BCL2\n",
      "  Model 1 Function: !BAD & STAT3\n",
      "  Model 2 Function: !BAD & MAPK3 & !MAPK8\n",
      "  Merged Function: !BAD & !MAPK8 & ( MAPK3 | STAT3 )\n",
      "\n",
      "Gene: CCND1\n",
      "  Model 1 Function: !GSK3B\n",
      "  Model 2 Function: !GSK3B & STAT3\n",
      "  Merged Function: !GSK3B & STAT3\n",
      "\n",
      "Gene: EGF\n",
      "  Model 1 Function: EGF\n",
      "  Model 2 Function: EGF\n",
      "  Merged Function: EGF\n",
      "\n",
      "Gene: EGFR\n",
      "  Model 1 Function: EGF\n",
      "  Model 2 Function: !MAPK3 & ( EGF | ERBB2 | HBEGF )\n",
      "  Merged Function: !MAPK3 & ( EGF | ERBB2 | HBEGF )\n",
      "\n",
      "Gene: ELK1\n",
      "  Model 1 Function: MAPK3 & RPS6KB1\n",
      "  Model 2 Function: MAPK3 | MAPK8\n",
      "  Merged Function: MAPK3 | MAPK8 | RPS6KB1\n",
      "\n",
      "Gene: ERBB2\n",
      "  Model 1 Function: NRG1\n",
      "  Model 2 Function: EGF | EGFR | NRG1\n",
      "  Merged Function: EGF | EGFR | NRG1\n",
      "\n",
      "Gene: ERBB4\n",
      "  Model 1 Function: EGF | HBEGF\n",
      "  Model 2 Function: !MAPK3 & ( ERBB2 | HBEGF | NRG1 )\n",
      "  Merged Function: !MAPK3 & ( EGF | ERBB2 | HBEGF | NRG1 )\n",
      "\n",
      "Gene: FOS\n",
      "  Model 1 Function: MAPK8 & RPS6KB1\n",
      "  Model 2 Function: MAPK3\n",
      "  Merged Function: MAPK3 | MAPK8 | RPS6KB1\n",
      "\n",
      "Gene: GRB2\n",
      "  Model 1 Function: EGFR | ERBB2 | ERBB4 | IGF1R\n",
      "  Model 2 Function: ERBB2 | ERBB4 | IRS1\n",
      "  Merged Function: EGFR | ERBB2 | ERBB4 | IGF1R | IRS1\n",
      "\n",
      "Gene: GSK3B\n",
      "  Model 1 Function: !AKT1\n",
      "  Model 2 Function: !AKT1\n",
      "  Merged Function: !AKT1\n",
      "\n",
      "Gene: HBEGF\n",
      "  Model 1 Function: HBEGF\n",
      "  Model 2 Function: HBEGF\n",
      "  Merged Function: HBEGF\n",
      "\n",
      "Gene: IGF1\n",
      "  Model 1 Function: IGF1\n",
      "  Model 2 Function: IGF1\n",
      "  Merged Function: IGF1\n",
      "\n",
      "Gene: IGF1R\n",
      "  Model 1 Function: IGF1\n",
      "  Model 2 Function: IGF1\n",
      "  Merged Function: IGF1\n",
      "\n",
      "Gene: IRS1\n",
      "  Model 1 Function: IGF1R\n",
      "  Model 2 Function: !MAPK3 & !MAPK8 & !MTOR & !PIK3CA & !PTEN & !RPS6KB1 & ( IGF1R | JAK1 )\n",
      "  Merged Function: !MAPK3 & !MAPK8 & !MTOR & !PIK3CA & !PTEN & !RPS6KB1 & ( IGF1R | JAK1 )\n",
      "\n",
      "Gene: JAK1\n",
      "  Model 1 Function: EGFR\n",
      "  Model 2 Function: EGFR\n",
      "  Merged Function: EGFR\n",
      "\n",
      "Gene: KRAS\n",
      "  Model 1 Function: GRB2\n",
      "  Model 2 Function: KRAS\n",
      "  Merged Function: GRB2 | KRAS\n",
      "\n",
      "Gene: MAP2K1\n",
      "  Model 1 Function: RAF1\n",
      "  Model 2 Function: MAP3K1 | RAF1\n",
      "  Merged Function: MAP3K1 | RAF1\n",
      "\n",
      "Gene: MAP2K4\n",
      "  Model 1 Function: MAP3K1\n",
      "  Model 2 Function: !AKT1 & MAP3K1\n",
      "  Merged Function: !AKT1 & MAP3K1\n",
      "\n",
      "Gene: MAP3K1\n",
      "  Model 1 Function: KRAS\n",
      "  Model 2 Function: MAP3K1\n",
      "  Merged Function: KRAS | MAP3K1\n",
      "\n",
      "Gene: MAPK3\n",
      "  Model 1 Function: MAP2K1\n",
      "  Model 2 Function: MAP2K1\n",
      "  Merged Function: MAP2K1\n",
      "\n",
      "Gene: MAPK8\n",
      "  Model 1 Function: MAP2K4\n",
      "  Model 2 Function: MAP2K4\n",
      "  Merged Function: MAP2K4\n",
      "\n",
      "Gene: MTOR\n",
      "  Model 1 Function: RHEB\n",
      "  Model 2 Function: PIK3CA & !RPS6KB1 & !TSC1\n",
      "  Merged Function: !RPS6KB1 & !TSC1 & ( PIK3CA | RHEB )\n",
      "\n",
      "Gene: NRG1\n",
      "  Model 1 Function: NRG1\n",
      "  Model 2 Function: NRG1\n",
      "  Merged Function: NRG1\n",
      "\n",
      "Gene: PDPK1\n",
      "  Model 1 Function: PIP3\n",
      "  Model 2 Function: !RPS6KB1\n",
      "  Merged Function: PIP3 & !RPS6KB1\n",
      "\n",
      "Gene: PIK3CA\n",
      "  Model 1 Function: ERBB2 | IRS1 | KRAS | STAT3\n",
      "  Model 2 Function: !PTEN & ( ERBB4 | IRS1 | KRAS )\n",
      "  Merged Function: !PTEN & ( ERBB2 | ERBB4 | IRS1 | KRAS | STAT3 )\n",
      "\n",
      "Gene: PRKAA1\n",
      "  Model 1 Function: STK11\n",
      "  Model 2 Function: STK11\n",
      "  Merged Function: STK11\n",
      "\n",
      "Gene: PTEN\n",
      "  Model 1 Function: PTEN\n",
      "  Model 2 Function: !STK11\n",
      "  Merged Function: PTEN & !STK11\n",
      "\n",
      "Gene: RAF1\n",
      "  Model 1 Function: KRAS\n",
      "  Model 2 Function: !AKT1 & !MAPK3 & ( KRAS | MAP2K1 )\n",
      "  Merged Function: !AKT1 & !MAPK3 & ( KRAS | MAP2K1 )\n",
      "\n",
      "Gene: RPS6KB1\n",
      "  Model 1 Function: MAPK3 | MTOR | PDPK1\n",
      "  Model 2 Function: !PTEN & ( MAPK3 | MTOR | PDPK1 )\n",
      "  Merged Function: !PTEN & ( MAPK3 | MTOR | PDPK1 )\n",
      "\n",
      "Gene: STAT3\n",
      "  Model 1 Function: JAK1\n",
      "  Model 2 Function: EGFR | JAK1 | MAPK3 | MAPK8 | MTOR\n",
      "  Merged Function: EGFR | JAK1 | MAPK3 | MAPK8 | MTOR\n",
      "\n",
      "Gene: STK11\n",
      "  Model 1 Function: STK11\n",
      "  Model 2 Function: STK11\n",
      "  Merged Function: STK11\n",
      "\n",
      "Gene: TSC1\n",
      "  Model 1 Function: !AKT1 | PRKAA1\n",
      "  Model 2 Function: TSC1\n",
      "  Merged Function: !AKT1 & ( PRKAA1 | TSC1 )\n"
     ]
    }
   ],
   "source": [
    "KG = BNMPy.load_network_from_string(bn_string)\n",
    "# Merge the networks using inhibitor wins\n",
    "merged_network_string = BNMPy.merge_networks([network, KG], method=\"Inhibitor Wins\", descriptive=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "69739b87-3bd3-4cc7-adf3-f6db2fac4375",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AKT1 = !PTEN & ( MTOR | PDPK1 | PIK3CA | PIP3 )\n",
      "BAD = !AKT1 & !MAPK8 & !RAF1 & RPS6KB1\n",
      "BCL2 = !BAD & !MAPK8 & ( MAPK3 | STAT3 )\n",
      "CCND1 = !GSK3B & STAT3\n",
      "EGF = EGF\n",
      "EGFR = !MAPK3 & ( EGF | ERBB2 | HBEGF )\n",
      "ELK1 = MAPK3 | MAPK8 | RPS6KB1\n",
      "ELK4 = MAPK3 & RPS6KB1\n",
      "ERBB2 = EGF | EGFR | NRG1\n",
      "ERBB4 = !MAPK3 & ( EGF | ERBB2 | HBEGF | NRG1 )\n",
      "FOS = MAPK3 | MAPK8 | RPS6KB1\n",
      "GRB2 = EGFR | ERBB2 | ERBB4 | IGF1R | IRS1\n",
      "GSK3B = !AKT1\n",
      "HBEGF = HBEGF\n",
      "IGF1 = IGF1\n",
      "IGF1R = IGF1\n",
      "IRS1 = !MAPK3 & !MAPK8 & !MTOR & !PIK3CA & !PTEN & !RPS6KB1 & ( IGF1R | JAK1 )\n",
      "JAK1 = EGFR\n",
      "KRAS = GRB2 | KRAS\n",
      "MAP2K1 = MAP3K1 | RAF1\n",
      "MAP2K4 = !AKT1 & MAP3K1\n",
      "MAP3K1 = KRAS | MAP3K1\n",
      "MAPK3 = MAP2K1\n",
      "MAPK8 = MAP2K4\n",
      "MTOR = !RPS6KB1 & !TSC1 & ( PIK3CA | RHEB )\n",
      "NRG1 = NRG1\n",
      "PDPK1 = PIP3 & !RPS6KB1\n",
      "PIK3CA = !PTEN & ( ERBB2 | ERBB4 | IRS1 | KRAS | STAT3 )\n",
      "PIP3 = PIK3CA | !PTEN\n",
      "PRKAA1 = STK11\n",
      "PTEN = PTEN & !STK11\n",
      "RAF1 = !AKT1 & !MAPK3 & ( KRAS | MAP2K1 )\n",
      "RHEB = !TSC1\n",
      "RPS6KB1 = !PTEN & ( MAPK3 | MTOR | PDPK1 )\n",
      "SP1 = MAPK3\n",
      "STAT3 = EGFR | JAK1 | MAPK3 | MAPK8 | MTOR\n",
      "STK11 = STK11\n",
      "TSC1 = !AKT1 & ( PRKAA1 | TSC1 )\n"
     ]
    }
   ],
   "source": [
    "print(merged_network_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9091319d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AKT1 = !PTEN & ( MTOR | PDPK1 | PIK3CA ), 0.1\n",
      "AKT1 = PIP3, 0.9\n",
      "BAD = ! ( AKT1 | RPS6KB1 ), 0.9\n",
      "BAD = !AKT1 & !MAPK8 & !RAF1, 0.1\n",
      "BCL2 = !BAD & MAPK3 & !MAPK8, 0.1\n",
      "BCL2 = !BAD & STAT3, 0.9\n",
      "CCND1 = !GSK3B & STAT3, 0.1\n",
      "CCND1 = !GSK3B, 0.9\n",
      "EGF = EGF, 0.9\n",
      "EGFR = !MAPK3 & ( EGF | ERBB2 | HBEGF ), 0.1\n",
      "EGFR = EGF, 0.9\n",
      "ELK1 = MAPK3 & RPS6KB1, 0.9\n",
      "ELK1 = MAPK3 | MAPK8, 0.1\n",
      "ELK4 = MAPK3 & RPS6KB1, 1.0\n",
      "ERBB2 = EGF | EGFR | NRG1, 0.1\n",
      "ERBB2 = NRG1, 0.9\n",
      "ERBB4 = !MAPK3 & ( ERBB2 | HBEGF | NRG1 ), 0.1\n",
      "ERBB4 = EGF | HBEGF, 0.9\n",
      "FOS = MAPK3, 0.1\n",
      "FOS = MAPK8 & RPS6KB1, 0.9\n",
      "GRB2 = EGFR | ERBB2 | ERBB4 | IGF1R, 0.9\n",
      "GRB2 = ERBB2 | ERBB4 | IRS1, 0.1\n",
      "GSK3B = !AKT1, 0.9\n",
      "HBEGF = HBEGF, 0.9\n",
      "IGF1 = IGF1, 0.9\n",
      "IGF1R = IGF1, 0.9\n",
      "IRS1 = !MAPK3 & !MAPK8 & !MTOR & !PIK3CA & !PTEN & !RPS6KB1 & ( IGF1R | JAK1 ), 0.1\n",
      "IRS1 = IGF1R, 0.9\n",
      "JAK1 = EGFR, 0.9\n",
      "KRAS = GRB2, 0.9\n",
      "KRAS = KRAS, 0.1\n",
      "MAP2K1 = MAP3K1 | RAF1, 0.1\n",
      "MAP2K1 = RAF1, 0.9\n",
      "MAP2K4 = !AKT1 & MAP3K1, 0.1\n",
      "MAP2K4 = MAP3K1, 0.9\n",
      "MAP3K1 = KRAS, 0.9\n",
      "MAP3K1 = MAP3K1, 0.1\n",
      "MAPK3 = MAP2K1, 0.9\n",
      "MAPK8 = MAP2K4, 0.9\n",
      "MTOR = PIK3CA & !RPS6KB1 & !TSC1, 0.1\n",
      "MTOR = RHEB, 0.9\n",
      "NRG1 = NRG1, 0.9\n",
      "PDPK1 = !RPS6KB1, 0.1\n",
      "PDPK1 = PIP3, 0.9\n",
      "PIK3CA = !PTEN & ( ERBB4 | IRS1 | KRAS ), 0.1\n",
      "PIK3CA = ERBB2 | IRS1 | KRAS | STAT3, 0.9\n",
      "PIP3 = PIK3CA | !PTEN, 1.0\n",
      "PRKAA1 = STK11, 0.9\n",
      "PTEN = !STK11, 0.1\n",
      "PTEN = PTEN, 0.9\n",
      "RAF1 = !AKT1 & !MAPK3 & ( KRAS | MAP2K1 ), 0.1\n",
      "RAF1 = KRAS, 0.9\n",
      "RHEB = !TSC1, 1.0\n",
      "RPS6KB1 = !PTEN & ( MAPK3 | MTOR | PDPK1 ), 0.1\n",
      "RPS6KB1 = MAPK3 | MTOR | PDPK1, 0.9\n",
      "SP1 = MAPK3, 1.0\n",
      "STAT3 = EGFR | JAK1 | MAPK3 | MAPK8 | MTOR, 0.1\n",
      "STAT3 = JAK1, 0.9\n",
      "STK11 = STK11, 0.9\n",
      "TSC1 = !AKT1 | PRKAA1, 0.9\n",
      "TSC1 = TSC1, 0.1\n"
     ]
    }
   ],
   "source": [
    "# Merge the networks using PBN\n",
    "# Here, a probability of 0.9 is used for rules from the original network\n",
    "merged_network_string = BNMPy.merge_networks([network, KG], method=\"PBN\", prob=0.9)\n",
    "print(merged_network_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0902514",
   "metadata": {},
   "source": [
    "## Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9a93062d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No initial state provided, using a random initial state\n",
      "PBN loaded successfully. There are 38 genes in the network.\n",
      "Network visualization saved to Vundavilli2020_extendedPBN.html\n"
     ]
    }
   ],
   "source": [
    "pbn = BNMPy.load_pbn_from_string(merged_network_string)\n",
    "BNMPy.vis_network(pbn, output_html=\"Vundavilli2020_extendedPBN.html\", interactive=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71888679",
   "metadata": {},
   "source": [
    "# Calculate phenotype score using KG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "618e8160",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 201 phenotypes\n",
      "There are 4905 genes\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "file_path = 'KG_files/significant_paths_to_phenotypes.txt'\n",
    "df = pd.read_csv(file_path, sep='\\t')\n",
    "print(f'There are {df[\"EndNode\"].nunique()} phenotypes')\n",
    "print(f'There are {df[\"QueryNode\"].nunique()} genes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "d0f46273",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ACROSOME_ASSEMBLY' 'ACTIN_CYTOSKELETON_REORGANIZATION'\n",
      " 'ACTION_POTENTIAL_' 'ADIPOGENESIS' 'ALTERNATIVE_SPLICING_REGULATION'\n",
      " 'AMYLOID_FIBRIL_FORMATION' 'ANGIOGENESIS' 'APOPTOSIS' 'ARDS'\n",
      " 'AUTOPHAGOSOME_FORMATION' 'AUTOPHAGY' 'AXONAL_GROWTH_CONE_FORMATION'\n",
      " 'B_CELL_MATURATION' 'B-LYMPHOCYTE_DIFF' 'BASOPHIL_DIFF'\n",
      " 'BONE_MINERALIZATION' 'BROWN_ADIPOGENESIS' 'CARTILAGE_DEVELOPMENT'\n",
      " 'CELL_ADHESION' 'CELL_CYCLE_BLOCK' 'CELL_CYCLE_EXIT'\n",
      " 'CELL_CYCLE_PROGRESS_' 'CELL_DEATH' 'CELL_GROWTH' 'CELL_KILLING'\n",
      " 'CELL_MIGRATION' 'CELL_POLARITY' 'CELL_SHAPE' 'CENTROMERE_ASSEMBLY'\n",
      " 'CENTROSOME_SEPARATION' 'CEREBRAL_CORTEX_DEVELOPMENT'\n",
      " 'CHAPERONE-MEDIATED_AUTOPHAGY' 'CHAPERONE-MEDIATED_PROTEIN_FOLDING'\n",
      " 'CHEMOATTRACTION_OF_AXON' 'CHEMOREPULSION_OF_AXON' 'CHEMOTAXIS'\n",
      " 'CHROMATINE_CONDENSATION' 'CHROMOSOME_SEGREGATION' 'CILIUM_ASSEMBLY'\n",
      " 'CILIUM_MOVEMENT' 'CITRIC_ACID_CYCLE'\n",
      " 'CLEARANCE_OF_FOREIGN_INTRACELLULAR_DNA' 'COLLOID' 'CYTOKINE_PRODUCTION'\n",
      " 'CYTOSKELETON_ORGANIZATION' 'CYTOTOXIC_T-LYMPHOCYTE_ACTIVATION'\n",
      " 'DEGRANULATION' 'DEMYELINATION' 'DENDRITIC_SPINE_MORPHOGENESIS'\n",
      " 'DENSE-CORE_VESICLE_EXOCYTOSIS' 'DIFFERENTIATION' 'DNA_FRAGMENTATION'\n",
      " 'DNA_REPAIR' 'DNA_REPLICATION' 'ECM_DISASSEMBLY' 'ECM_SYNTHESIS'\n",
      " 'ENDOCYTOSIS' 'EPIGENETIC_REGULATION' 'EPITHELIAL-MESENCHYMAL_TRANSITION'\n",
      " 'ERYTHROCYTE_DIFFERENTIATION_' 'EXCITATORY_SYNAPTIC_TRANSMISSION'\n",
      " 'EXOCYTOSIS' 'F-ACTIN_ASSEMBLY' 'FATTY_ACID_BIOSYNTHESIS'\n",
      " 'FATTY_ACID_OXIDATION' 'FIBRIN_CLOT_FORMATION' 'FIBRINOLYSIS' 'FIBROSIS'\n",
      " 'FOOD_INTAKE' 'G1/S_TRANSITION' 'G1/S_TRANSITION_CHECKPOINT'\n",
      " 'G2/M_TRANSITION' 'G2/M_TRANSITION-CHECKPOINT' 'GLUCONEOGENESIS'\n",
      " 'GLUTAMINOLYSIS' 'GLYCOGEN_SYNTHESIS' 'GLYCOLYSIS'\n",
      " 'GRANULOCYTE_DIFFERENTIATION' 'HDL_ASSEMBLY'\n",
      " 'HELPER_T-LYMPHOCYTE_ACTIVATION' 'HEXOSAMINE_BIOSYNTHESIS' 'IMMORTALITY'\n",
      " 'IMMUNE_RESPONSE' 'INFLAMMATION' 'INSERTION_INTO_MITOCHONDRIAL_MEMBRANE'\n",
      " 'INTERFERON_PRODUCTION' 'LDL_ASSEMBLY' 'LEWY_BODY_FORMATION'\n",
      " 'LIPOGENESIS' 'LYMPHOMA' 'LYMPHOPOIESIS' 'M1_POLARIZATION'\n",
      " 'M2_POLARIZATION' 'MACROPHAGE_ACTIVATION' 'MACROPHAGE_DIFFERENTIATION'\n",
      " 'MAINTENANCE_OF_LENS_TRANSPARENCY' 'MAST-CELL_DIFF'\n",
      " 'MEGAKARYOCYTE_DIFFERENTIATION_' 'MEIOTIC_RECOMBINATION'\n",
      " 'MELANOSOME_ASSEMBLY' 'MEMBRANE_BLEBBING' 'MEMBRANE_DISRUPTION'\n",
      " 'MEMBRANE_FUSION' 'METABOLISM' 'METASTASIS' 'MICROTUBULE_POLIMERIZATION'\n",
      " 'MICROTUBULE-BASED_MOVEMENT' 'MINUS-END_DIRECTED_MICROTUBULE_MOVEMENT'\n",
      " 'MITOCHONDRIAL_BIOGENESIS' 'MITOCHONDRIAL_FISSION' 'MITOTIC_CHECKPOINT'\n",
      " 'MONOCYTE_DIFFERENTIATION_' 'MRNA_CAPPING' 'MRNA_POLYADENYLATION'\n",
      " 'MRNA_SCANNING' 'MRNA-NUCLEUS_EXPORT' 'MULTIVESICULAR_BODY_ASSEMBLY'\n",
      " 'MUSCLE_ATROPHY' 'MYELINATION' 'MYOBLAST_FUSION' 'MYOTUBE_HYPERTROPHY'\n",
      " 'NCRNA_PROCESSING' 'NECROPTOSIS' 'NECROSIS' 'NEURAL_TISSUE_REGENERATION'\n",
      " 'NEURITE_OUTGROWTH' 'NEURODEGENERATION'\n",
      " 'NEUROFIBRILLARY_TANGLE_FORMATION' 'NEUROFILAMENT_BUNDLE_ASSEMBLY'\n",
      " 'NEUROGENESIS' 'NEURON_MATURATION' 'NEURON_MIGRATION'\n",
      " 'NEUTROPHIL_ACTIVATION' 'NONSENSE-MEDIATED_MRNA_DECAY'\n",
      " 'NUCLEAR_PORE_FUNCTION' 'NUCLEOTIDE_SYNTHESIS'\n",
      " 'NUCLEOTIDE-EXCISION_REPAIR' 'ORGANELLE_TRANSPORT' 'OSSIFICATION'\n",
      " 'OSTEOBLAST_DIFFERENTIATION' 'OSTEOCLAST_DIFFERENTIATION'\n",
      " 'OXIDATIVE_PHOSPHORYLATION' 'OXIDATIVE_STRESS'\n",
      " 'PEPTIDE_HORMONE_PROCESSING' 'PHAGOCYTOSIS' 'PIGMENTATION'\n",
      " 'PLATELET_ADHESION_' 'PLATELET_AGGREGATION'\n",
      " 'PLATELET_ALPHA_GRANULE_FORMATION' 'PLATELET_DEGRANULATION'\n",
      " 'PLATELET_DENSE_GRANULE_FORMATION' 'PLATELET_MORPHOGENESIS'\n",
      " 'PLURIPOTENCY' 'PLUS-END_DIRECTED_SLIDING_MOVEMENT'\n",
      " 'POSTSYNAPTIC_DENSITY_ASSEMBLY' 'PROLIFERATION' 'PROTEIN_ACETYLATION'\n",
      " 'PROTEIN_AGGREGATES' 'PROTEIN_DEGRADATION' 'PROTEIN_GLYCOSYLATION'\n",
      " 'PROTEIN_LOCALIZATION_TO_PEROXISOME' 'PROTEIN_SYNTHESIS'\n",
      " 'PROTEIN_TRANSLOCATION' 'PROTEIN_UBIQUITINATION' 'PURINE_BIOSYNTHESIS'\n",
      " 'PYRIMIDINE_BIOSYNTHESIS' 'PYRIMIDINE_NUCLEOTIDE_METABOLIC_PROCESS'\n",
      " 'PYROPTOSIS' 'QUIESCENCE' 'RECEPTOR_MEDIATED__ENDOCYTOSIS'\n",
      " 'RESPIRATORY_ELECTRON_TRANSPORT_CHAIN' 'RIBOSOME_BIOGENESIS'\n",
      " 'RNA_SPLICING' 'RRNA_TRANSCRIPTION' 'SARCOMERE_ORGANIZATION'\n",
      " 'SATELLITE_CELLS_SELF-RENEWAL' 'SECRETORY_GRANULE_ORGANIZATION'\n",
      " 'SKELETAL_MUSCLE_DIFFERENTIATION' 'SPINDLE_ASSEMBLY'\n",
      " 'SPLICEOSOMAL_SNRNP_ASSEMBLY' 'STRESS_GRANULES' 'SURVIVAL'\n",
      " 'SYNAPTIC_PLASTICITY' 'SYNAPTIC_VESICLE_EXOCYTOSIS'\n",
      " 'SYNAPTIC_VESICLE_RECYCLING' 'T_CELL_ACTIVATION' 'T-LYMPHOCYTE_DIFF'\n",
      " 'T-REG_DIFFERENTIATION' 'TELOMERE_MAINTENANCE' 'TH17' 'THERMOGENESIS_'\n",
      " 'THYROID_HORMONOGENESIS' 'TRANSCRITPIONAL_ACTIVATION'\n",
      " 'TRANSLATIONAL_ELONGATION' 'TRANSLATIONAL_REGULATION' 'UPR'\n",
      " 'UROGENITAL_TRACT' 'VASCULAR_PERMEABILITY' 'VESICLE_TRANSPORT'\n",
      " 'VIRAL_BUDDING' 'VLDL_ASSEMBLY']\n"
     ]
    }
   ],
   "source": [
    "print(df[\"EndNode\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bdc9688",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EndPathways</th>\n",
       "      <th>QueryNode</th>\n",
       "      <th>EndNode</th>\n",
       "      <th>Path_String</th>\n",
       "      <th>relations_path</th>\n",
       "      <th>Path_Score</th>\n",
       "      <th>Path_Length</th>\n",
       "      <th>Final_Effect</th>\n",
       "      <th>Effect</th>\n",
       "      <th>n</th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>zscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>53815</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>CDKN2A--|CDK4--[]CYCLIND/CDK4--|RB1--|PROLIFER...</td>\n",
       "      <td>SIGNOR-44554;SIGNOR-32301;SIGNOR-250762;SIGNOR...</td>\n",
       "      <td>0.577</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.182991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53913</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>CDKN2A--|PROLIFERATION</td>\n",
       "      <td>SIGNOR-259406</td>\n",
       "      <td>0.300</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.730415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54033</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>CDKN2A--|CYCLIND/CDK4--|RB1--|PROLIFERATION</td>\n",
       "      <td>SIGNOR-245459;SIGNOR-250762;SIGNOR-262533</td>\n",
       "      <td>0.633</td>\n",
       "      <td>3</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.072320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54152</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>CDKN2A--|CDK6--|RB1--|PROLIFERATION</td>\n",
       "      <td>SIGNOR-44557;SIGNOR-135189;SIGNOR-262533</td>\n",
       "      <td>0.672</td>\n",
       "      <td>3</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-1.995246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54306</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>CDKN2A--|CDK4--|RB1--|PROLIFERATION</td>\n",
       "      <td>SIGNOR-44554;SIGNOR-200483;SIGNOR-262533</td>\n",
       "      <td>0.464</td>\n",
       "      <td>3</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.406308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54518</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>CDKN2A--|CDK6--|CDKN1A--|CYCLINE/CDK2--&gt;PROLIF...</td>\n",
       "      <td>SIGNOR-44557;SIGNOR-144832;SIGNOR-245462;SIGNO...</td>\n",
       "      <td>0.651</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.036748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54124</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>KRAS--&gt;PIK3CA--[]PI3K--&gt;PROLIFERATION</td>\n",
       "      <td>SIGNOR-175204;SIGNOR-255299;SIGNOR-255577</td>\n",
       "      <td>0.450</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>up-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.433976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54127</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>KRAS--&gt;PIK3CA--&gt;AKT--&gt;PROLIFERATION</td>\n",
       "      <td>SIGNOR-175204;SIGNOR-244429;SIGNOR-254353</td>\n",
       "      <td>0.600</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>up-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.137537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54568</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>KRAS--&gt;PIK3CA--[]PI3K--&gt;AKT--&gt;PROLIFERATION</td>\n",
       "      <td>SIGNOR-175204;SIGNOR-255299;SIGNOR-254950;SIGN...</td>\n",
       "      <td>0.681</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>up-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-1.977460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54569</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>KRAS--&gt;PIK3CA--[]PI3K--&gt;AKT1--&gt;PROLIFERATION</td>\n",
       "      <td>SIGNOR-175204;SIGNOR-255299;SIGNOR-255106;SIGN...</td>\n",
       "      <td>0.681</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>up-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-1.977460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54600</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>KRAS--&gt;PI3K--&gt;PROLIFERATION</td>\n",
       "      <td>SIGNOR-252698;SIGNOR-255577</td>\n",
       "      <td>0.574</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>up-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.188920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53515</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>TP53</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>TP53--&gt;CDKN1A--|CYCLINE/CDK2--&gt;PROLIFERATION</td>\n",
       "      <td>SIGNOR-173425;SIGNOR-245462;SIGNOR-262532</td>\n",
       "      <td>0.536</td>\n",
       "      <td>3</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.264018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53529</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>TP53</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>TP53--|PROLIFERATION</td>\n",
       "      <td>SIGNOR-255669</td>\n",
       "      <td>0.300</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.730415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53553</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>TP53</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>TP53--&gt;CDKN1A--|CDK4--|RB1--|PROLIFERATION</td>\n",
       "      <td>SIGNOR-173425;SIGNOR-29957;SIGNOR-200483;SIGNO...</td>\n",
       "      <td>0.552</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.232398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53556</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>TP53</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>TP53--&gt;CDKN1A--|CDK2--[]CYCLINE/CDK2--&gt;PROLIFE...</td>\n",
       "      <td>SIGNOR-173425;SIGNOR-149711;SIGNOR-187457;SIGN...</td>\n",
       "      <td>0.537</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.262041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53557</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>TP53</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>TP53--&gt;CDKN1A--|CDK2--|RB1--|PROLIFERATION</td>\n",
       "      <td>SIGNOR-173425;SIGNOR-149711;SIGNOR-47895;SIGNO...</td>\n",
       "      <td>0.579</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.179038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54592</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>TP53</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>TP53--&gt;CDKN1A--|CDK2--|CDKN1B--|PROLIFERATION</td>\n",
       "      <td>SIGNOR-173425;SIGNOR-149711;SIGNOR-154188;SIGN...</td>\n",
       "      <td>0.524</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.287733</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         EndPathways QueryNode        EndNode  \\\n",
       "53815  PROLIFERATION    CDKN2A  PROLIFERATION   \n",
       "53913  PROLIFERATION    CDKN2A  PROLIFERATION   \n",
       "54033  PROLIFERATION    CDKN2A  PROLIFERATION   \n",
       "54152  PROLIFERATION    CDKN2A  PROLIFERATION   \n",
       "54306  PROLIFERATION    CDKN2A  PROLIFERATION   \n",
       "54518  PROLIFERATION    CDKN2A  PROLIFERATION   \n",
       "54124  PROLIFERATION      KRAS  PROLIFERATION   \n",
       "54127  PROLIFERATION      KRAS  PROLIFERATION   \n",
       "54568  PROLIFERATION      KRAS  PROLIFERATION   \n",
       "54569  PROLIFERATION      KRAS  PROLIFERATION   \n",
       "54600  PROLIFERATION      KRAS  PROLIFERATION   \n",
       "53515  PROLIFERATION      TP53  PROLIFERATION   \n",
       "53529  PROLIFERATION      TP53  PROLIFERATION   \n",
       "53553  PROLIFERATION      TP53  PROLIFERATION   \n",
       "53556  PROLIFERATION      TP53  PROLIFERATION   \n",
       "53557  PROLIFERATION      TP53  PROLIFERATION   \n",
       "54592  PROLIFERATION      TP53  PROLIFERATION   \n",
       "\n",
       "                                             Path_String  \\\n",
       "53815  CDKN2A--|CDK4--[]CYCLIND/CDK4--|RB1--|PROLIFER...   \n",
       "53913                             CDKN2A--|PROLIFERATION   \n",
       "54033        CDKN2A--|CYCLIND/CDK4--|RB1--|PROLIFERATION   \n",
       "54152                CDKN2A--|CDK6--|RB1--|PROLIFERATION   \n",
       "54306                CDKN2A--|CDK4--|RB1--|PROLIFERATION   \n",
       "54518  CDKN2A--|CDK6--|CDKN1A--|CYCLINE/CDK2-->PROLIF...   \n",
       "54124              KRAS-->PIK3CA--[]PI3K-->PROLIFERATION   \n",
       "54127                KRAS-->PIK3CA-->AKT-->PROLIFERATION   \n",
       "54568        KRAS-->PIK3CA--[]PI3K-->AKT-->PROLIFERATION   \n",
       "54569       KRAS-->PIK3CA--[]PI3K-->AKT1-->PROLIFERATION   \n",
       "54600                        KRAS-->PI3K-->PROLIFERATION   \n",
       "53515       TP53-->CDKN1A--|CYCLINE/CDK2-->PROLIFERATION   \n",
       "53529                               TP53--|PROLIFERATION   \n",
       "53553         TP53-->CDKN1A--|CDK4--|RB1--|PROLIFERATION   \n",
       "53556  TP53-->CDKN1A--|CDK2--[]CYCLINE/CDK2-->PROLIFE...   \n",
       "53557         TP53-->CDKN1A--|CDK2--|RB1--|PROLIFERATION   \n",
       "54592      TP53-->CDKN1A--|CDK2--|CDKN1B--|PROLIFERATION   \n",
       "\n",
       "                                          relations_path  Path_Score  \\\n",
       "53815  SIGNOR-44554;SIGNOR-32301;SIGNOR-250762;SIGNOR...       0.577   \n",
       "53913                                      SIGNOR-259406       0.300   \n",
       "54033          SIGNOR-245459;SIGNOR-250762;SIGNOR-262533       0.633   \n",
       "54152           SIGNOR-44557;SIGNOR-135189;SIGNOR-262533       0.672   \n",
       "54306           SIGNOR-44554;SIGNOR-200483;SIGNOR-262533       0.464   \n",
       "54518  SIGNOR-44557;SIGNOR-144832;SIGNOR-245462;SIGNO...       0.651   \n",
       "54124          SIGNOR-175204;SIGNOR-255299;SIGNOR-255577       0.450   \n",
       "54127          SIGNOR-175204;SIGNOR-244429;SIGNOR-254353       0.600   \n",
       "54568  SIGNOR-175204;SIGNOR-255299;SIGNOR-254950;SIGN...       0.681   \n",
       "54569  SIGNOR-175204;SIGNOR-255299;SIGNOR-255106;SIGN...       0.681   \n",
       "54600                        SIGNOR-252698;SIGNOR-255577       0.574   \n",
       "53515          SIGNOR-173425;SIGNOR-245462;SIGNOR-262532       0.536   \n",
       "53529                                      SIGNOR-255669       0.300   \n",
       "53553  SIGNOR-173425;SIGNOR-29957;SIGNOR-200483;SIGNO...       0.552   \n",
       "53556  SIGNOR-173425;SIGNOR-149711;SIGNOR-187457;SIGN...       0.537   \n",
       "53557  SIGNOR-173425;SIGNOR-149711;SIGNOR-47895;SIGNO...       0.579   \n",
       "54592  SIGNOR-173425;SIGNOR-149711;SIGNOR-154188;SIGN...       0.524   \n",
       "\n",
       "       Path_Length  Final_Effect          Effect       n      mean        sd  \\\n",
       "53815            4            -1  down-regulates  107507  1.681607  0.506006   \n",
       "53913            1            -1  down-regulates  107507  1.681607  0.506006   \n",
       "54033            3            -1  down-regulates  107507  1.681607  0.506006   \n",
       "54152            3            -1  down-regulates  107507  1.681607  0.506006   \n",
       "54306            3            -1  down-regulates  107507  1.681607  0.506006   \n",
       "54518            4            -1  down-regulates  107507  1.681607  0.506006   \n",
       "54124            3             1    up-regulates  107507  1.681607  0.506006   \n",
       "54127            3             1    up-regulates  107507  1.681607  0.506006   \n",
       "54568            4             1    up-regulates  107507  1.681607  0.506006   \n",
       "54569            4             1    up-regulates  107507  1.681607  0.506006   \n",
       "54600            2             1    up-regulates  107507  1.681607  0.506006   \n",
       "53515            3            -1  down-regulates  107507  1.681607  0.506006   \n",
       "53529            1            -1  down-regulates  107507  1.681607  0.506006   \n",
       "53553            4            -1  down-regulates  107507  1.681607  0.506006   \n",
       "53556            4            -1  down-regulates  107507  1.681607  0.506006   \n",
       "53557            4            -1  down-regulates  107507  1.681607  0.506006   \n",
       "54592            4            -1  down-regulates  107507  1.681607  0.506006   \n",
       "\n",
       "         zscore  \n",
       "53815 -2.182991  \n",
       "53913 -2.730415  \n",
       "54033 -2.072320  \n",
       "54152 -1.995246  \n",
       "54306 -2.406308  \n",
       "54518 -2.036748  \n",
       "54124 -2.433976  \n",
       "54127 -2.137537  \n",
       "54568 -1.977460  \n",
       "54569 -1.977460  \n",
       "54600 -2.188920  \n",
       "53515 -2.264018  \n",
       "53529 -2.730415  \n",
       "53553 -2.232398  \n",
       "53556 -2.262041  \n",
       "53557 -2.179038  \n",
       "54592 -2.287733  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def proxpath(genes, phenotypes = ['APOPTOSIS', 'DIFFERENTIATION', 'PROLIFERATION'], file_path = 'KG_files/significant_paths_to_phenotypes.txt'):\n",
    "    # Load the ProxPath file\n",
    "    df = pd.read_csv(file_path, sep='\\t')\n",
    "    \n",
    "    # Function to find the closest gene to the EndNode\n",
    "    def closest_gene(path_string, genes):\n",
    "        # Split the path into components and reverse it (to start from the EndNode)\n",
    "        components = path_string.split('--')[::-1]\n",
    "        for component in components:\n",
    "            # Check if the component contains any of the genes\n",
    "            for gene in genes:\n",
    "                if gene in component:\n",
    "                    return gene\n",
    "        return None\n",
    "    \n",
    "    # Filter rows for the given genes and phenotypes\n",
    "    filtered_df = df[df['QueryNode'].isin(genes) & df['EndNode'].isin(phenotypes)].copy()\n",
    "    \n",
    "    # find the closest gene to the phenotype\n",
    "    filtered_df['Closest_Gene'] = filtered_df.apply(lambda row: closest_gene(row['Path_String'], genes), axis=1)\n",
    "    \n",
    "    # Filter rows where QueryNode is the closest gene to the phenotype\n",
    "    pheno_df = filtered_df[filtered_df['QueryNode'] == filtered_df['Closest_Gene']]\n",
    "    pheno_df = pheno_df.drop(columns=['Closest_Gene'])\n",
    "\n",
    "    # sort\n",
    "    pheno_df = pheno_df.sort_values(by=['EndNode', 'QueryNode'])\n",
    "\n",
    "    # remove rows where Final_Effect is 0\n",
    "    pheno_df = pheno_df[pheno_df['Final_Effect'] != 0]\n",
    "    \n",
    "    # save the filtered data to a new file\n",
    "    # pheno_df.to_csv('Phenotypes.txt', sep='\\t', index=False)\n",
    "\n",
    "    return pheno_df\n",
    "\n",
    "pheno_df = proxpath(genes = ['KRAS', 'GNAS', 'TP53', 'SMAD4', 'CDKN2A', 'RNF43'], phenotypes = ['PROLIFERATION'])\n",
    "pheno_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f90732fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EndPathways</th>\n",
       "      <th>QueryNode</th>\n",
       "      <th>EndNode</th>\n",
       "      <th>Path_String</th>\n",
       "      <th>relations_path</th>\n",
       "      <th>Path_Score</th>\n",
       "      <th>Path_Length</th>\n",
       "      <th>Final_Effect</th>\n",
       "      <th>Effect</th>\n",
       "      <th>n</th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>zscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>53913</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>CDKN2A--|PROLIFERATION</td>\n",
       "      <td>SIGNOR-259406</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.730415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54124</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>KRAS--&gt;PIK3CA--[]PI3K--&gt;PROLIFERATION</td>\n",
       "      <td>SIGNOR-175204;SIGNOR-255299;SIGNOR-255577</td>\n",
       "      <td>0.45</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>up-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.433976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53529</th>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>TP53</td>\n",
       "      <td>PROLIFERATION</td>\n",
       "      <td>TP53--|PROLIFERATION</td>\n",
       "      <td>SIGNOR-255669</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>107507</td>\n",
       "      <td>1.681607</td>\n",
       "      <td>0.506006</td>\n",
       "      <td>-2.730415</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         EndPathways QueryNode        EndNode  \\\n",
       "53913  PROLIFERATION    CDKN2A  PROLIFERATION   \n",
       "54124  PROLIFERATION      KRAS  PROLIFERATION   \n",
       "53529  PROLIFERATION      TP53  PROLIFERATION   \n",
       "\n",
       "                                 Path_String  \\\n",
       "53913                 CDKN2A--|PROLIFERATION   \n",
       "54124  KRAS-->PIK3CA--[]PI3K-->PROLIFERATION   \n",
       "53529                   TP53--|PROLIFERATION   \n",
       "\n",
       "                                  relations_path  Path_Score  Path_Length  \\\n",
       "53913                              SIGNOR-259406        0.30            1   \n",
       "54124  SIGNOR-175204;SIGNOR-255299;SIGNOR-255577        0.45            3   \n",
       "53529                              SIGNOR-255669        0.30            1   \n",
       "\n",
       "       Final_Effect          Effect       n      mean        sd    zscore  \n",
       "53913            -1  down-regulates  107507  1.681607  0.506006 -2.730415  \n",
       "54124             1    up-regulates  107507  1.681607  0.506006 -2.433976  \n",
       "53529            -1  down-regulates  107507  1.681607  0.506006 -2.730415  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# there may be different Final_Effect values for the same QueryNode and EndNode\n",
    "for gene in pheno_df['QueryNode'].unique():\n",
    "    for phenotype in pheno_df['EndNode'].unique():\n",
    "        if pheno_df[(pheno_df['QueryNode'] == gene) & (pheno_df['EndNode'] == phenotype)]['Final_Effect'].nunique() > 1:\n",
    "            print(f\"{gene} has dual effects on {phenotype}\")\n",
    "\n",
    "# Keep only the rows with the lowest Path_Score for each gene-phenotype pair\n",
    "pheno_df.loc[pheno_df.groupby(['QueryNode', 'EndNode'])['Path_Score'].idxmin()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "33af653b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EndPathways</th>\n",
       "      <th>QueryNode</th>\n",
       "      <th>EndNode</th>\n",
       "      <th>Path_String</th>\n",
       "      <th>relations_path</th>\n",
       "      <th>Path_Score</th>\n",
       "      <th>Path_Length</th>\n",
       "      <th>Final_Effect</th>\n",
       "      <th>Effect</th>\n",
       "      <th>n</th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>zscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2903</th>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>BAX</td>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>BAX--&gt;APOPTOSIS</td>\n",
       "      <td>SIGNOR-261494</td>\n",
       "      <td>0.300</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>up-regulates</td>\n",
       "      <td>75493</td>\n",
       "      <td>1.737625</td>\n",
       "      <td>0.526944</td>\n",
       "      <td>-2.728233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2913</th>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>BCL2</td>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>BCL2--|APOPTOSIS</td>\n",
       "      <td>SIGNOR-249611</td>\n",
       "      <td>0.300</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>75493</td>\n",
       "      <td>1.737625</td>\n",
       "      <td>0.526944</td>\n",
       "      <td>-2.728233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3437</th>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>BCL2</td>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>BCL2--|BAK1--&gt;APOPTOSIS</td>\n",
       "      <td>SIGNOR-152980;SIGNOR-261493</td>\n",
       "      <td>0.589</td>\n",
       "      <td>2</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>75493</td>\n",
       "      <td>1.737625</td>\n",
       "      <td>0.526944</td>\n",
       "      <td>-2.179787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2948</th>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>KRAS--&gt;PIK3CA--&gt;AKT--|FOXO--&gt;APOPTOSIS</td>\n",
       "      <td>SIGNOR-175204;SIGNOR-244429;SIGNOR-252824;SIGN...</td>\n",
       "      <td>0.698</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>75493</td>\n",
       "      <td>1.737625</td>\n",
       "      <td>0.526944</td>\n",
       "      <td>-1.972934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3343</th>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>KRAS--&gt;PIK3CA--[]PI3K--&gt;AKT--|APOPTOSIS</td>\n",
       "      <td>SIGNOR-175204;SIGNOR-255299;SIGNOR-254950;SIGN...</td>\n",
       "      <td>0.681</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>75493</td>\n",
       "      <td>1.737625</td>\n",
       "      <td>0.526944</td>\n",
       "      <td>-2.005196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3539</th>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>KRAS--&gt;PIK3CA--&gt;AKT--|APOPTOSIS</td>\n",
       "      <td>SIGNOR-175204;SIGNOR-244429;SIGNOR-260215</td>\n",
       "      <td>0.600</td>\n",
       "      <td>3</td>\n",
       "      <td>-1</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>75493</td>\n",
       "      <td>1.737625</td>\n",
       "      <td>0.526944</td>\n",
       "      <td>-2.158912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2894</th>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>TP53</td>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>TP53--&gt;APOPTOSIS</td>\n",
       "      <td>SIGNOR-255678</td>\n",
       "      <td>0.300</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>up-regulates</td>\n",
       "      <td>75493</td>\n",
       "      <td>1.737625</td>\n",
       "      <td>0.526944</td>\n",
       "      <td>-2.728233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3195</th>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>TP53</td>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>TP53--&gt;BAK1--&gt;APOPTOSIS</td>\n",
       "      <td>SIGNOR-124122;SIGNOR-261493</td>\n",
       "      <td>0.615</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>up-regulates</td>\n",
       "      <td>75493</td>\n",
       "      <td>1.737625</td>\n",
       "      <td>0.526944</td>\n",
       "      <td>-2.130446</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     EndPathways QueryNode    EndNode  \\\n",
       "2903   APOPTOSIS       BAX  APOPTOSIS   \n",
       "2913   APOPTOSIS      BCL2  APOPTOSIS   \n",
       "3437   APOPTOSIS      BCL2  APOPTOSIS   \n",
       "2948   APOPTOSIS      KRAS  APOPTOSIS   \n",
       "3343   APOPTOSIS      KRAS  APOPTOSIS   \n",
       "3539   APOPTOSIS      KRAS  APOPTOSIS   \n",
       "2894   APOPTOSIS      TP53  APOPTOSIS   \n",
       "3195   APOPTOSIS      TP53  APOPTOSIS   \n",
       "\n",
       "                                  Path_String  \\\n",
       "2903                          BAX-->APOPTOSIS   \n",
       "2913                         BCL2--|APOPTOSIS   \n",
       "3437                  BCL2--|BAK1-->APOPTOSIS   \n",
       "2948   KRAS-->PIK3CA-->AKT--|FOXO-->APOPTOSIS   \n",
       "3343  KRAS-->PIK3CA--[]PI3K-->AKT--|APOPTOSIS   \n",
       "3539          KRAS-->PIK3CA-->AKT--|APOPTOSIS   \n",
       "2894                         TP53-->APOPTOSIS   \n",
       "3195                  TP53-->BAK1-->APOPTOSIS   \n",
       "\n",
       "                                         relations_path  Path_Score  \\\n",
       "2903                                      SIGNOR-261494       0.300   \n",
       "2913                                      SIGNOR-249611       0.300   \n",
       "3437                        SIGNOR-152980;SIGNOR-261493       0.589   \n",
       "2948  SIGNOR-175204;SIGNOR-244429;SIGNOR-252824;SIGN...       0.698   \n",
       "3343  SIGNOR-175204;SIGNOR-255299;SIGNOR-254950;SIGN...       0.681   \n",
       "3539          SIGNOR-175204;SIGNOR-244429;SIGNOR-260215       0.600   \n",
       "2894                                      SIGNOR-255678       0.300   \n",
       "3195                        SIGNOR-124122;SIGNOR-261493       0.615   \n",
       "\n",
       "      Path_Length  Final_Effect          Effect      n      mean        sd  \\\n",
       "2903            1             1    up-regulates  75493  1.737625  0.526944   \n",
       "2913            1            -1  down-regulates  75493  1.737625  0.526944   \n",
       "3437            2            -1  down-regulates  75493  1.737625  0.526944   \n",
       "2948            4            -1  down-regulates  75493  1.737625  0.526944   \n",
       "3343            4            -1  down-regulates  75493  1.737625  0.526944   \n",
       "3539            3            -1  down-regulates  75493  1.737625  0.526944   \n",
       "2894            1             1    up-regulates  75493  1.737625  0.526944   \n",
       "3195            2             1    up-regulates  75493  1.737625  0.526944   \n",
       "\n",
       "        zscore  \n",
       "2903 -2.728233  \n",
       "2913 -2.728233  \n",
       "3437 -2.179787  \n",
       "2948 -1.972934  \n",
       "3343 -2.005196  \n",
       "3539 -2.158912  \n",
       "2894 -2.728233  \n",
       "3195 -2.130446  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pheno_df = proxpath(genes = ['TP53', 'KRAS', 'MYC', 'SMAD4', 'BCL2', 'BAX'], phenotypes = ['APOPTOSIS'])\n",
    "pheno_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7afd71db",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1177386",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pheno_scores(simulation_results, pheno):\n",
    "    \n",
    "    # Keep only the rows with the lowest Path_Score for each gene-phenotype pair\n",
    "    pheno_unique = pheno.loc[pheno.groupby(['QueryNode', 'EndNode'])['Path_Score'].idxmin()]\n",
    "\n",
    "    phenotypes = ['APOPTOSIS', 'DIFFERENTIATION', 'PROLIFERATION']    \n",
    "    # Loop through each phenotype\n",
    "    for phenotype in phenotypes:\n",
    "        \n",
    "        # Filter the 'pheno' dataframe for the current phenotype\n",
    "        filtered_pheno = pheno_unique[pheno_unique['EndNode'] == phenotype]\n",
    "\n",
    "        # Loop through each row in the filtered 'pheno' dataframe\n",
    "        for idx, row in filtered_pheno.iterrows():\n",
    "            gene = row['QueryNode']\n",
    "            effect = row['Final_Effect']\n",
    "            \n",
    "            # Check if the gene is in the simulation results\n",
    "            if gene in simulation_results.index:\n",
    "                # Multiply the gene's simulation result by its effect (1 or -1)\n",
    "                simulation_results.loc[phenotype] += simulation_results.loc[gene] * effect\n",
    "\n",
    "    return simulation_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4f3b36a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 201 phenotypes\n",
      "There are 4905 genes\n",
      "Available phenotypes: ['ACROSOME_ASSEMBLY' 'ACTIN_CYTOSKELETON_REORGANIZATION'\n",
      " 'ACTION_POTENTIAL_' 'ADIPOGENESIS' 'ALTERNATIVE_SPLICING_REGULATION'\n",
      " 'AMYLOID_FIBRIL_FORMATION' 'ANGIOGENESIS' 'APOPTOSIS' 'ARDS'\n",
      " 'AUTOPHAGOSOME_FORMATION' 'AUTOPHAGY' 'AXONAL_GROWTH_CONE_FORMATION'\n",
      " 'B_CELL_MATURATION' 'B-LYMPHOCYTE_DIFF' 'BASOPHIL_DIFF'\n",
      " 'BONE_MINERALIZATION' 'BROWN_ADIPOGENESIS' 'CARTILAGE_DEVELOPMENT'\n",
      " 'CELL_ADHESION' 'CELL_CYCLE_BLOCK' 'CELL_CYCLE_EXIT'\n",
      " 'CELL_CYCLE_PROGRESS_' 'CELL_DEATH' 'CELL_GROWTH' 'CELL_KILLING'\n",
      " 'CELL_MIGRATION' 'CELL_POLARITY' 'CELL_SHAPE' 'CENTROMERE_ASSEMBLY'\n",
      " 'CENTROSOME_SEPARATION' 'CEREBRAL_CORTEX_DEVELOPMENT'\n",
      " 'CHAPERONE-MEDIATED_AUTOPHAGY' 'CHAPERONE-MEDIATED_PROTEIN_FOLDING'\n",
      " 'CHEMOATTRACTION_OF_AXON' 'CHEMOREPULSION_OF_AXON' 'CHEMOTAXIS'\n",
      " 'CHROMATINE_CONDENSATION' 'CHROMOSOME_SEGREGATION' 'CILIUM_ASSEMBLY'\n",
      " 'CILIUM_MOVEMENT' 'CITRIC_ACID_CYCLE'\n",
      " 'CLEARANCE_OF_FOREIGN_INTRACELLULAR_DNA' 'COLLOID' 'CYTOKINE_PRODUCTION'\n",
      " 'CYTOSKELETON_ORGANIZATION' 'CYTOTOXIC_T-LYMPHOCYTE_ACTIVATION'\n",
      " 'DEGRANULATION' 'DEMYELINATION' 'DENDRITIC_SPINE_MORPHOGENESIS'\n",
      " 'DENSE-CORE_VESICLE_EXOCYTOSIS' 'DIFFERENTIATION' 'DNA_FRAGMENTATION'\n",
      " 'DNA_REPAIR' 'DNA_REPLICATION' 'ECM_DISASSEMBLY' 'ECM_SYNTHESIS'\n",
      " 'ENDOCYTOSIS' 'EPIGENETIC_REGULATION' 'EPITHELIAL-MESENCHYMAL_TRANSITION'\n",
      " 'ERYTHROCYTE_DIFFERENTIATION_' 'EXCITATORY_SYNAPTIC_TRANSMISSION'\n",
      " 'EXOCYTOSIS' 'F-ACTIN_ASSEMBLY' 'FATTY_ACID_BIOSYNTHESIS'\n",
      " 'FATTY_ACID_OXIDATION' 'FIBRIN_CLOT_FORMATION' 'FIBRINOLYSIS' 'FIBROSIS'\n",
      " 'FOOD_INTAKE' 'G1/S_TRANSITION' 'G1/S_TRANSITION_CHECKPOINT'\n",
      " 'G2/M_TRANSITION' 'G2/M_TRANSITION-CHECKPOINT' 'GLUCONEOGENESIS'\n",
      " 'GLUTAMINOLYSIS' 'GLYCOGEN_SYNTHESIS' 'GLYCOLYSIS'\n",
      " 'GRANULOCYTE_DIFFERENTIATION' 'HDL_ASSEMBLY'\n",
      " 'HELPER_T-LYMPHOCYTE_ACTIVATION' 'HEXOSAMINE_BIOSYNTHESIS' 'IMMORTALITY'\n",
      " 'IMMUNE_RESPONSE' 'INFLAMMATION' 'INSERTION_INTO_MITOCHONDRIAL_MEMBRANE'\n",
      " 'INTERFERON_PRODUCTION' 'LDL_ASSEMBLY' 'LEWY_BODY_FORMATION'\n",
      " 'LIPOGENESIS' 'LYMPHOMA' 'LYMPHOPOIESIS' 'M1_POLARIZATION'\n",
      " 'M2_POLARIZATION' 'MACROPHAGE_ACTIVATION' 'MACROPHAGE_DIFFERENTIATION'\n",
      " 'MAINTENANCE_OF_LENS_TRANSPARENCY' 'MAST-CELL_DIFF'\n",
      " 'MEGAKARYOCYTE_DIFFERENTIATION_' 'MEIOTIC_RECOMBINATION'\n",
      " 'MELANOSOME_ASSEMBLY' 'MEMBRANE_BLEBBING' 'MEMBRANE_DISRUPTION'\n",
      " 'MEMBRANE_FUSION' 'METABOLISM' 'METASTASIS' 'MICROTUBULE_POLIMERIZATION'\n",
      " 'MICROTUBULE-BASED_MOVEMENT' 'MINUS-END_DIRECTED_MICROTUBULE_MOVEMENT'\n",
      " 'MITOCHONDRIAL_BIOGENESIS' 'MITOCHONDRIAL_FISSION' 'MITOTIC_CHECKPOINT'\n",
      " 'MONOCYTE_DIFFERENTIATION_' 'MRNA_CAPPING' 'MRNA_POLYADENYLATION'\n",
      " 'MRNA_SCANNING' 'MRNA-NUCLEUS_EXPORT' 'MULTIVESICULAR_BODY_ASSEMBLY'\n",
      " 'MUSCLE_ATROPHY' 'MYELINATION' 'MYOBLAST_FUSION' 'MYOTUBE_HYPERTROPHY'\n",
      " 'NCRNA_PROCESSING' 'NECROPTOSIS' 'NECROSIS' 'NEURAL_TISSUE_REGENERATION'\n",
      " 'NEURITE_OUTGROWTH' 'NEURODEGENERATION'\n",
      " 'NEUROFIBRILLARY_TANGLE_FORMATION' 'NEUROFILAMENT_BUNDLE_ASSEMBLY'\n",
      " 'NEUROGENESIS' 'NEURON_MATURATION' 'NEURON_MIGRATION'\n",
      " 'NEUTROPHIL_ACTIVATION' 'NONSENSE-MEDIATED_MRNA_DECAY'\n",
      " 'NUCLEAR_PORE_FUNCTION' 'NUCLEOTIDE_SYNTHESIS'\n",
      " 'NUCLEOTIDE-EXCISION_REPAIR' 'ORGANELLE_TRANSPORT' 'OSSIFICATION'\n",
      " 'OSTEOBLAST_DIFFERENTIATION' 'OSTEOCLAST_DIFFERENTIATION'\n",
      " 'OXIDATIVE_PHOSPHORYLATION' 'OXIDATIVE_STRESS'\n",
      " 'PEPTIDE_HORMONE_PROCESSING' 'PHAGOCYTOSIS' 'PIGMENTATION'\n",
      " 'PLATELET_ADHESION_' 'PLATELET_AGGREGATION'\n",
      " 'PLATELET_ALPHA_GRANULE_FORMATION' 'PLATELET_DEGRANULATION'\n",
      " 'PLATELET_DENSE_GRANULE_FORMATION' 'PLATELET_MORPHOGENESIS'\n",
      " 'PLURIPOTENCY' 'PLUS-END_DIRECTED_SLIDING_MOVEMENT'\n",
      " 'POSTSYNAPTIC_DENSITY_ASSEMBLY' 'PROLIFERATION' 'PROTEIN_ACETYLATION'\n",
      " 'PROTEIN_AGGREGATES' 'PROTEIN_DEGRADATION' 'PROTEIN_GLYCOSYLATION'\n",
      " 'PROTEIN_LOCALIZATION_TO_PEROXISOME' 'PROTEIN_SYNTHESIS'\n",
      " 'PROTEIN_TRANSLOCATION' 'PROTEIN_UBIQUITINATION' 'PURINE_BIOSYNTHESIS'\n",
      " 'PYRIMIDINE_BIOSYNTHESIS' 'PYRIMIDINE_NUCLEOTIDE_METABOLIC_PROCESS'\n",
      " 'PYROPTOSIS' 'QUIESCENCE' 'RECEPTOR_MEDIATED__ENDOCYTOSIS'\n",
      " 'RESPIRATORY_ELECTRON_TRANSPORT_CHAIN' 'RIBOSOME_BIOGENESIS'\n",
      " 'RNA_SPLICING' 'RRNA_TRANSCRIPTION' 'SARCOMERE_ORGANIZATION'\n",
      " 'SATELLITE_CELLS_SELF-RENEWAL' 'SECRETORY_GRANULE_ORGANIZATION'\n",
      " 'SKELETAL_MUSCLE_DIFFERENTIATION' 'SPINDLE_ASSEMBLY'\n",
      " 'SPLICEOSOMAL_SNRNP_ASSEMBLY' 'STRESS_GRANULES' 'SURVIVAL'\n",
      " 'SYNAPTIC_PLASTICITY' 'SYNAPTIC_VESICLE_EXOCYTOSIS'\n",
      " 'SYNAPTIC_VESICLE_RECYCLING' 'T_CELL_ACTIVATION' 'T-LYMPHOCYTE_DIFF'\n",
      " 'T-REG_DIFFERENTIATION' 'TELOMERE_MAINTENANCE' 'TH17' 'THERMOGENESIS_'\n",
      " 'THYROID_HORMONOGENESIS' 'TRANSCRITPIONAL_ACTIVATION'\n",
      " 'TRANSLATIONAL_ELONGATION' 'TRANSLATIONAL_REGULATION' 'UPR'\n",
      " 'UROGENITAL_TRACT' 'VASCULAR_PERMEABILITY' 'VESICLE_TRANSPORT'\n",
      " 'VIRAL_BUDDING' 'VLDL_ASSEMBLY']\n"
     ]
    }
   ],
   "source": [
    "from BNMPy.phenotype_score import phenotype_scores, get_phenotypes\n",
    "get_phenotypes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "074ad947",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Path found for 3 phenotypes: ['APOPTOSIS' 'PROLIFERATION' 'DIFFERENTIATION']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'APOPTOSIS': 'BAX -  BCL2 -  KRAS + TP53',\n",
       " 'PROLIFERATION': 'KRAS + MYC -  TP53',\n",
       " 'DIFFERENTIATION': 'MYC + TP53'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenotype_scores(\n",
    "    genes = ['TP53', 'KRAS', 'MYC', 'SMAD4', 'BCL2', 'BAX'], \n",
    "    phenotypes = ['APOPTOSIS', 'DIFFERENTIATION', 'PROLIFERATION'], \n",
    "    file_path = 'KG_files/significant_paths_to_phenotypes.txt', \n",
    "    simulation_results=None\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "061f0987",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "998e5efd",
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_scores(\n",
    "    genes = ['TP53', 'KRAS', 'MYC', 'SMAD4', 'BCL2', 'BAX'], \n",
    "    phenotypes = ['APOPTOSIS', 'DIFFERENTIATION', 'PROLIFERATION'], \n",
    "    file_path = 'KG_files/significant_paths_to_phenotypes.txt', \n",
    "    simulation_results=None\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5fba2d4",
   "metadata": {},
   "source": [
    "# Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d46a5537",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['subject_id', 'object_id', 'subject_id_prefix', 'object_id_prefix',\n",
      "       'subject_name', 'object_name', 'predicate', 'Primary_Knowledge_Source',\n",
      "       'Knowledge_Source', 'publications', 'subject_category',\n",
      "       'object_category', 'score'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "signor_file = 'KG_files/SIGNOR_2025_08_14.tsv'\n",
    "graph_table = pd.read_csv(signor_file, index_col=None, sep='\\t')\n",
    "print(graph_table.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5ce0a15e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    40935.000000\n",
       "mean         0.501978\n",
       "std          0.235996\n",
       "min          0.100000\n",
       "25%          0.278000\n",
       "50%          0.468000\n",
       "75%          0.727000\n",
       "max          1.000000\n",
       "Name: score, dtype: float64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_table['score'].describe()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "logicmodelmerger",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
